api/services/CasebookService.js

```diff
@@ -1,53 +1,65 @@
 const crypto = require('crypto');
-const apiQueryString = require('querystring');
-const request = require('request');
-const requestPromise = require('request-promise');
-const sails = require('sails');
-
-const CasebookService = {
-    get(options) {
-        return CasebookService._returnRequestMethods('get', options);
-    },
+const axios = require('axios');
+const https = require('https');
+
+const config = require('../../config/environment-variables')
+const {
+    hmacKey,
+    casebookCertificate: cert,
+    casebookKey: key,
+    customURLs,
+} = config;
 
-    post(options) {
-        return CasebookService._returnRequestMethods('post', options);
+const CasebookService = axios.create({
+    baseURL: customURLs.casebookBaseUrl,
+    headers: {
+        'Content-Type': 'application/json; charset=utf-8',
+        'api-version': '4',
     },
+    httpsAgent: new https.Agent({
+        cert,
+        key,
+        keepAlive: true,
+    }),
+    paramsSerializer: queryParamObjToStr,
+    transformRequest: [addHmacToQueryParam],
+});
 
-    _returnRequestMethods(method, options) {
-        const requestType = options.promise ? requestPromise : request;
-        delete options.promise;
+function queryParamObjToStr(queryParamsObj) {
+    const params = new URLSearchParams();
 
-        const authParams = CasebookService._createAuthParams(options.qs);
-        const optionsWithAuthParams = { ...options, ...authParams };
+    Object.entries(queryParamsObj).forEach(([key, value]) => {
+        if (Array.isArray(value)) {
+            for (const val of value) {
+                params.append(key, val);
+            }
+        } else {
+            params.append(key, value.toString());
+        }
+    });
 
-        return requestType[method](optionsWithAuthParams);
-    },
+    return params.toString();
+}
 
-    _createAuthParams(queryParamsObj) {
-        const {
-            hmacKey,
-            casebookCertificate: cert,
-            casebookKey: key,
-        } = sails.config;
-
-        const queryStr = apiQueryString.stringify(queryParamsObj);
-
-        const hash = crypto
-            .createHmac('sha512', hmacKey)
-            .update(Buffer.from(queryStr, 'utf-8'))
-            .digest('hex')
-            .toUpperCase();
-
-        return {
-            agentOptions: {
-                cert,
-                key,
-            },
-            headers: {
-                hash,
-            },
-        };
-    },
-};
+function addHmacToQueryParam(data, _headers) {
+    data = {
+        hmac: hmacKey
+    };
+    return JSON.stringify(data);
+}
+
+CasebookService.interceptors.request.use(addHashToHeader);
```

Comment:
That would be ideal but I couldn't figure out how to do it. For the has to work it would need access to the params and I didn't know the best way to add arguments to `axios.create`.

Also, I thought the request intercept would happen before the request is made? 🤔

Reply:
should've been clearer, sorry, so with intercept what happens is call is made -> axios intercepts it and changes -> request carries on to wherever it's heading 

but ideally it should be request called with everything it needs without the interception

Reply:
in the docs/comments for transformRequest
```
  // You may modify the headers object.
  transformRequest: [function (data, headers) {
```

if you can't get it working intercept will be ok

Reply:
I've tried it without intercept but by passing an argument in. Let me know your thoughts on that approach.

Reply:
In response to this question (#305 (comment)) transformRequest only applies to 'PUT', 'POST', 'PATCH' and 'DELETE' and no 'GET' 😢 so this won't quite work.